File: homework/week2/circular_queue.py
MAX_SIZE = 100
class Queue:
    def __init__(self):
        self.list = [None] * MAX_SIZE
        self.front = 0
        self.rear = 0
        self.count = 0

q = Queue()

def isFull(q):
    return q.count == MAX_SIZE

def isEmpty(q):
    return q.count == 0

def enqueue(q, element):
    if isFull(q):
        return None
    q.list[q.rear] = element
    q.rear = (q.rear + 1) % MAX_SIZE
    q.count += 1

def dequeue(q):
    if isEmpty(q):
        return
    value = q.list[q.front]
    q.front = (q.front + 1) % MAX_SIZE
    q.count -= 1
    return value
    

File: homework/week2/2025320058-twuel0921/CircularQueue.java
public class CircularQueue<T> {
    private final int MAX_SIZE = 100;
    private int front;
    private int rear;
    private int size;
    private Object[] elemetArray;

    public CircularQueue() {
        elemetArray = new Object[MAX_SIZE];
        front = 0;
        rear = 0;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == MAX_SIZE;
    }

    public int size() {
        return size;
    }

    public T enqueue(T item) {
        if(isFull()) {
            throw new ArrayIndexOutOfBoundsException("Full Queue!");
        }
        elemetArray[rear] = item;
        rear = (rear+1)%MAX_SIZE;
        size++;
        return item;
    }

    public T dequeue() {
        if(isEmpty()) {
            throw new ArrayIndexOutOfBoundsException("Empty Queue!");
        }
        T item = (T) elemetArray[front];
        front = (front+1)%MAX_SIZE;
        size--;
        return item;
    }
}
